<a href="https://colab.research.google.com/github/Anupwilson/MS-in-Data-Analytics-DBS-assignments/blob/main/CA2_Programming_for_Data_analytics__Rakesh_Nishanth_Anup.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Colab Link: https://colab.research.google.com/drive/1VBDVHY69nomzIU6Mn1V3IBO8MvRDXa3N#scrollTo=_GlS28I_GaZo

WEB SCRAPING OF **FLIPKART** **WEBSITE** FOR MOBILE PHONES UNDER 50000 RANGE

Name:**Rakesh Vemula**

Student Id:**10641241**

It is necessary to choose proper ecommerce real-time website for our assignment task in order to perform Webscrapping. **Flipkart website** was selected after extensive research. Website contains multiple categories 
in that we have chosen **mobile phones** category. Each product contains multiple features in that we have chosen **Product Name**, **Price**, **Description**, **No.of ratings** and **Reviews** for our data analysis.

Scrapping of data ,acquiring of data and to store the acquired data we need to import some python libraries. We have imported pandas, requests and BeautipulSoup libraries.

Processing html documents we have **html.parser** and **lxml**
here I have chosen **lxml** parser because it is a better choice for working with large complex html documents where as **html.parser** is better choice for doing small projects.

Dynamically scrapped the data from 70 pages, looping through each page and stored the corresponding data into their lists. Exceptions are also handled in the code by using **try** and **except** blocks.

Used **from_dist** function from the pandas dataframe to convert python dictionary into dataframe and also used **transpose** function because dictionary keys are formed as rows and index's as columns.

Imported the webscrapped data into csv file.

We imported prepocessed data into csv file.Then after i have imported the csv file into mongodb server.

I was able to gain more knowledge regarding the webscrapping, usage of imported packages, converting python dictionary into pandas dataframe, creating csv file from dataframe, reading of data from csv file and inserting into mongodb server.

---------------------------------------------------------------------------
Name: **NISHANTH HARISH**

Student Id: **20000522**

The assignment involved series of processes, starting from web-scraping, data pre-processing, importing data to a database, developing a data analysis pipeline, and answering various questions from the dataset.

 

My keen interest was in data pre-processing stage, as I was really into data cleaning, removing duplicate or irrelevant observations, identifying and filtering unwanted outliers, handling missing data.

 

During brainstorming sessions with teammates, we came up with a road map on how we can answer interesting questions. Keeping that in mind, we have specifically identified the Product Name, Price, Description, No. of Ratings, and Reviews as the key variables of interest. These variables will undergo data pre-processing to ensure that they are formatted correctly and are free of any errors or inconsistencies. By cleaning and standardizing the data, we can generate accurate and reliable insights that can be used to inform our research conclusions.

1. While working in the pre-processing stage, I firstly imported necessary libraries and loaded the dataset that was obtained after webscraping.

2. Next, I tried to check the number of observations present in the dataset by using describe(). This gave me the description of reviews like count, mean, maximum, minimum values, standard deviation.

3. I checked if there any null values and dropped the null values

4. In the middle of this, I wanted to know how many unique values were present in the Reviews column and tried printing them so that this would help in answering a question.

5. Further I created bins to accommodate the reviews with unique values to sit in particular labels (Low, Medium, High) and created a new column called Label_Reviews.

6. Since the new column was successfully loaded, I dropped the Reviews column to avoid confusion by keeping only Label_Reviews column and printed the dataframe.

7. In this step I again repeated step 4 but now with newly created Label_Reviews column and Prices column to find unique values. In this output it was difficult to accommodate all the Prices in our analysis hence I created another dataframe as Label_prices and defined the categories, so that certain range of products based on price falls under particular category (Like budget phones, smart phones, premium phones).

8. After consulting with my teammates, this dataset was finalised to load into to Mongodb. Hence, I again cleaned the dataset and made it ready for the data analysis part and also converted this preprocessed data into CSV file.

I had previously worked on data processing on multiple occasions but had never tried webscraping or accessing a Mongodb. So I was eager in engaging with other steps with my friends to try and understand the process involved. Also while coming up with thoughtful questions, I had proposed an idea but it required a data modelling. Overall, it was a wonderful learning curve in this course as this assignment required critical thinking and analysis on the real time data.

---------------------------------------------------------------------------
Name: **ANUP WILSON**

Student Id: **10628141**

In this assignment, our primary objective was to overcome multiple challenges and implement a robust data analysis pipeline involving web scraping, data preprocessing, and analysis. Our approach involved formulating critical questions that were significant for both us and our customers purchasing the products. The primary focus was on key parameters affecting the product's quality and customer satisfaction. my primary interest was to design a data analysis pipeline that could deliver meaningful insights and help us make data-driven decisions. As a result, we adopted a professional approach to address the challenges and achieve our objective of delivering valuable insights to our customers.

1. Initially, we faced several challenges in determining the appropriate website for data scraping. After careful consideration, we decided to use amazon.co.uk for our data collection needs. However, we encountered a significant obstacle during the scraping process, as we discovered that the website's base URL would change periodically for approximately 20-30 minutes. To resolve this issue, Vemula Rakesh, who had previous experience and expertise in web scraping, took on the responsibility of finding a solution. He explored alternative websites and eventually identified flipkart.com as a viable option. Subsequently, we collaborated with Vemula Rakesh to gain a better understanding of web scraping techniques. As a result, we acquired valuable insights into the workings of web scraping and its various nuances.

2. After selecting the appropriate website for web scraping, we proceeded with data collection and converted the data into CSV format for further processing. Given my extensive experience in the domain, I took on the responsibility of developing the data analysis pipeline for this assignment, as I believed it to be a critical component of the project. Meanwhile, Nishanth Harish undertook the task of preprocessing the raw data and provided me with a clean and structured dataset that met my specific requirements. Specifically, I requested him to build bins for certain attributes that I deemed crucial for creating an effective data analysis pipeline. Nishanth Harish utilized his own techniques and expertise to deliver a meaningful dataset that aligned with my project objectives.

3. After receiving the finalized pre-processed data, we uploaded it to the MongoDB server with the name "flipkart" as per our project requirements. My primary responsibility was to design insightful questions based on my domain expertise and knowledge that could potentially aid customers in their product purchasing decisions.

4. Following the completion of the project, I was able to develop approximately 9-10 meaningful questions based on my subject expertise. However, with further reflection, I recognized that my initial focus was not on queries that could potentially aid customers in their purchasing decisions. Instead, I was drawn to the insights and patterns that could be gleaned from studying customer purchase behavior. This insight sparked my interest and inspired me to study the amount of knowledge contained in the data analysis pipeline in further detail.

5. Following the formulation of pertinent questions, I embarked on developing a data analysis pipeline that leveraged various techniques such as logical operators in Python. Additionally, I utilized the Natural Language Toolkit (NLTK), a machine learning algorithm, to extract the most common features present in the product descriptions. Through the implementation of these tools and techniques, I was able to generate valuable insights and make informed decisions based on the analyzed data. The collaborative effort of our team was crucial in ensuring the success of the project.

Overall, the project presented numerous challenges that required our team to work collaboratively and draw upon our collective expertise in order to overcome them. Some of the challenges we faced included identifying an appropriate website for web scraping, navigating the intricacies of data preprocessing, and designing insightful questions that could potentially benefit customers in their purchasing decisions.

Throughout the project, I learned the importance of effective communication and collaboration among team members, as well as the significance of leveraging appropriate tools and techniques to generate meaningful insights from data. Additionally, I gained valuable experience in developing data analysis pipelines, which can be utilized in future projects to derive insights from large datasets.

Despite these challenges, our team was able to successfully complete the project and derive valuable insights from the analyzed data.




---------------------------------------------------------------------------




**Importing the required python packages or libraries in order to perform web scrapping**

In [1]:
import pandas as pd
import requests
from bs4 import BeautifulSoup

**Webscrapping and creating of dataset code goes here**

In [2]:
#lists to store data
Product_names=[]
Prices=[]
Description=[]
Reviews=[]
No_of_Ratings=[]

for i in range(2,70):
    url="https://www.flipkart.com/search?q=mobiles+under+50000&otracker=search&otracker1=search&marketplace=FLIPKART&as-show=off&as=off&as-pos=1&as-type=HISTORY&page="+str(i)
    page=requests.get(url)
    soup=BeautifulSoup(page.text,'lxml')
    box=soup.find("div",class_="_1YokD2 _3Mn1Gg")
    names=box.find_all("div",class_="_4rR01T")
    for n in names:
        name=n.text
        try:
          Product_names.append(name)
        except:
            Product_names.append(None)
    #print(Product_names)
    prices=box.find_all("div",class_="_30jeq3 _1_WHN1") 
    for p in prices:
        price=p.text.split('₹')[1]
        try:
          Prices.append(price)
        except:
            Prices.append(None)
    #print(Prices)
    descp=box.find_all("div",class_="fMghEO")
    for d in descp:
        des=d.text
        try:
          Description.append(des)
        except:
            Description.append(None)
    #print(Description)
    revs=box.find_all("div",class_="_3LWZlK")
    for r in revs:
        review=r.text
        try:
            Reviews.append(review)
        except:
            Reviews.append(None)
    #print(Reviews)
    noofRatings=box.find_all("span",class_="_2_R_DZ")
    for r in noofRatings:
        rating=r.text.split(' ')[0]
        try:
            No_of_Ratings.append(rating)
        except:
            No_of_Ratings.append(None)
    #print(No_of_Ratings)
    
productData = {"Product Name":Product_names,"Prices":Prices,"Description":Description,"Reviews":Reviews,"No. of Ratings": No_of_Ratings}
df=pd.DataFrame.from_dict(productData,orient="index")
df=df.transpose()
#df=pd.DataFrame({"Product Name":Product_names,"Prices":Prices,"Description":Description,"Reviews":Reviews,"No of Ratings": No_of_Ratings})
print(df)
df.head()

#while True:
   # np=soup.find("a",class_="_1LKTO3").get("href")
    #print(np)
    #cnp="https://www.flipkart.com"+np
    #print(cnp)
    #url=cnp
    #page=requests.get(url)
    #soup=BeautifulSoup(page.text,'html.parser')

                                          Product Name  Prices  \
0                 realme X50 Pro 5G (Rust Red, 256 GB)  47,999   
1              Xiaomi 12 Pro 5G (Couture Blue, 256 GB)  48,990   
2                Xiaomi 12 Pro 5G (Noir Black, 256 GB)  45,890   
3                                        Good One 5026   1,099   
4              APPLE iPhone 8 Plus (Space Grey, 64 GB)  49,900   
..                                                 ...     ...   
955                                         I Kall K78   1,204   
956               POCO M4 Pro 5G (Power Black, 128 GB)  15,999   
957                    POCO X4 Pro 5G (Yellow, 128 GB)  19,999   
958  Lava Pulse1- with contactless Talking Thermometer   1,460   
959                 REDMI 10A SPORT (SEA BLUE, 128 GB)  10,280   

                                           Description Reviews No. of Ratings  
0    12 GB RAM | 256 GB ROM16.36 cm (6.44 inch) Ful...     4.4            778  
1    12 GB RAM | 256 GB ROM17.09 cm (6.73 inch)

Product Name  Prices  \
0     realme X50 Pro 5G (Rust Red, 256 GB)  47,999   
1  Xiaomi 12 Pro 5G (Couture Blue, 256 GB)  48,990   
2    Xiaomi 12 Pro 5G (Noir Black, 256 GB)  45,890   
3                            Good One 5026   1,099   
4  APPLE iPhone 8 Plus (Space Grey, 64 GB)  49,900   

                                         Description Reviews No. of Ratings  
0  12 GB RAM | 256 GB ROM16.36 cm (6.44 inch) Ful...     4.4            778  
1  12 GB RAM | 256 GB ROM17.09 cm (6.73 inch) Ful...     4.2            177  
2  8 GB RAM | 256 GB ROM17.09 cm (6.73 inch) Full...     4.1            184  
3  32 MB RAM | 32 MB ROM | Expandable Upto 32 GB4...     3.9          1,127  
4  64 GB ROM13.97 cm (5.5 inch) Retina HD Display...     4.6          3,438

Dynamic web scrapping was built to acquire the data from 70 pages in the Mobile phones category.


The information we have obtained includes the Product Name, Prices, Description, No.of Ratings and Reviews.

The Python script listed above web scrapes the Flipkart website to gather information about cell phones priced under 50,000 rupees. A for loop is used to extract the data from several pages, and for each page, the script gathers the product name, price, description, customer reviews, and number of ratings for each mobile phone.

The script parses the HTML text of each webpage and extracts the necessary data using the BeautifulSoup library. The information is then kept in distinct lists for each category and appended to each scraped page.

The extracted data is then turned into a Pandas DataFrame, which is then printed to the console. The script also contains commented-out code that may be used to scrape more data from each cell phone's product page, such as the features and specifications.


**Import the Web Scraped Data into csv file**

In [3]:
df.to_csv("flipcartwebscrapeddata.csv",index=False)

Here in this code the scrapped data from the flipkart website is coverted into csv file format and the same as been saved.

# Data Preprocessing **Stage**

**Importing libraries required to perfrom data preproccessing such as Pandas, Numpy, Matplotlib, Etc.,**



In [4]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

**Importing of Datset:** Loading of dataset we have obtained from the webscraping for Data preprocessing




In [5]:
df = pd.read_csv("flipcartwebscrapeddata.csv")

In [6]:
df.count()

Product Name      960
Prices            960
Description       960
Reviews           931
No. of Ratings    931
dtype: int64

The above codes provides the total no of records present in the dataset that we have scrapped.

In [7]:
df.isnull().sum()

Product Name       0
Prices             0
Description        0
Reviews           29
No. of Ratings    29
dtype: int64

This code is used to find out the total number null records present in each attributes present in the dataset

In [8]:
df.isna().sum()

Product Name       0
Prices             0
Description        0
Reviews           29
No. of Ratings    29
dtype: int64

This code is used to find out the total number na records present in each attributes present in the dataset

In [9]:
df.describe()

Reviews
count  931.000000
mean     4.168636
std      0.239767
min      3.300000
25%      4.000000
50%      4.200000
75%      4.300000
max      4.700000

The above code is used to find basic stats of the numerical columns present in the dataset.

In [10]:
!hostname -i

172.28.0.12


In [11]:
df=df.dropna()

Using the above code we have dropped all the na values prsent in the dataset.

In [12]:
df.count()

Product Name      931
Prices            931
Description       931
Reviews           931
No. of Ratings    931
dtype: int64

Here we can see that the dataset had total of 960 records in Product Name,
Prices, Description and 931 records in Reviews, No. of Ratings before droping the na or null values, once we have dropped the null or na values the data contains total of 931 records in each attributes.

In [13]:
df.isnull().sum()

Product Name      0
Prices            0
Description       0
Reviews           0
No. of Ratings    0
dtype: int64

In [14]:
df.isnull().sum()

Product Name      0
Prices            0
Description       0
Reviews           0
No. of Ratings    0
dtype: int64

we are re-checking for the null or nan values in the dataset post dropping of the nan or null values if any.

In [15]:
df.describe()

Reviews
count  931.000000
mean     4.168636
std      0.239767
min      3.300000
25%      4.000000
50%      4.200000
75%      4.300000
max      4.700000

In [16]:
df.head()

Product Name  Prices  \
0     realme X50 Pro 5G (Rust Red, 256 GB)  47,999   
1  Xiaomi 12 Pro 5G (Couture Blue, 256 GB)  48,990   
2    Xiaomi 12 Pro 5G (Noir Black, 256 GB)  45,890   
3                            Good One 5026   1,099   
4  APPLE iPhone 8 Plus (Space Grey, 64 GB)  49,900   

                                         Description  Reviews No. of Ratings  
0  12 GB RAM | 256 GB ROM16.36 cm (6.44 inch) Ful...      4.4            778  
1  12 GB RAM | 256 GB ROM17.09 cm (6.73 inch) Ful...      4.2            177  
2  8 GB RAM | 256 GB ROM17.09 cm (6.73 inch) Full...      4.1            184  
3  32 MB RAM | 32 MB ROM | Expandable Upto 32 GB4...      3.9          1,127  
4  64 GB ROM13.97 cm (5.5 inch) Retina HD Display...      4.6          3,438

In [17]:
print(df['Reviews'].unique())

[4.4 4.2 4.1 3.9 4.6 4.  3.5 4.3 4.5 3.8 4.7 3.6 3.4 3.7 3.3]


In [18]:
bins = ['1', '2.5', '3.5', '5']
labels = ['Low', 'Medium', 'High']
df['Label_Reviews'] = pd.cut(df['Reviews'], bins=bins, labels=labels)

Using the above code we have created a bins for the Reviews attribute.

**Bins are created as mentioned below:**

Reviews laying between **1-2.5** are mentioned has **Low**.

Reviews laying between **2.5-3.5** are mentioned has **Medium**.

Reviews laying between **3.5-5** are mentioned has **High**.

And a seperate dummy column has been created to check whether the bins were properly implemented or not.

In [19]:
df.head()

Product Name  Prices  \
0     realme X50 Pro 5G (Rust Red, 256 GB)  47,999   
1  Xiaomi 12 Pro 5G (Couture Blue, 256 GB)  48,990   
2    Xiaomi 12 Pro 5G (Noir Black, 256 GB)  45,890   
3                            Good One 5026   1,099   
4  APPLE iPhone 8 Plus (Space Grey, 64 GB)  49,900   

                                         Description  Reviews No. of Ratings  \
0  12 GB RAM | 256 GB ROM16.36 cm (6.44 inch) Ful...      4.4            778   
1  12 GB RAM | 256 GB ROM17.09 cm (6.73 inch) Ful...      4.2            177   
2  8 GB RAM | 256 GB ROM17.09 cm (6.73 inch) Full...      4.1            184   
3  32 MB RAM | 32 MB ROM | Expandable Upto 32 GB4...      3.9          1,127   
4  64 GB ROM13.97 cm (5.5 inch) Retina HD Display...      4.6          3,438   

  Label_Reviews  
0          High  
1          High  
2          High  
3          High  
4          High

In [20]:
df = df.drop('Reviews', axis=1)

Here we have droped the Reviews atribute keeping the proccessed attribute of Reviews.

In [21]:
df.head()

Product Name  Prices  \
0     realme X50 Pro 5G (Rust Red, 256 GB)  47,999   
1  Xiaomi 12 Pro 5G (Couture Blue, 256 GB)  48,990   
2    Xiaomi 12 Pro 5G (Noir Black, 256 GB)  45,890   
3                            Good One 5026   1,099   
4  APPLE iPhone 8 Plus (Space Grey, 64 GB)  49,900   

                                         Description No. of Ratings  \
0  12 GB RAM | 256 GB ROM16.36 cm (6.44 inch) Ful...            778   
1  12 GB RAM | 256 GB ROM17.09 cm (6.73 inch) Ful...            177   
2  8 GB RAM | 256 GB ROM17.09 cm (6.73 inch) Full...            184   
3  32 MB RAM | 32 MB ROM | Expandable Upto 32 GB4...          1,127   
4  64 GB ROM13.97 cm (5.5 inch) Retina HD Display...          3,438   

  Label_Reviews  
0          High  
1          High  
2          High  
3          High  
4          High

In [22]:
print(df['Label_Reviews'].unique())

['High', 'Medium']
Categories (3, object): ['Low' < 'Medium' < 'High']


In [23]:
print(df['Prices'].unique())

['47,999' '48,990' '45,890' '1,099' '49,900' '49,499' '45,999' '899'
 '44,999' '44,399' '34,999' '40,999' '36,890' '42,999' '29,990' '41,999'
 '41,990' '43,499' '36,999' '39,842' '31,988' '30,999' '39,999' '29,999'
 '37,999' '38,999' '38,990' '35,999' '36,990' '31,978' '37,949' '35,990'
 '26,399' '39,900' '46,599' '49,999' '42,021' '36,630' '39,990' '36,790'
 '36,299' '37,990' '9,999' '10,999' '11,999' '12,999' '7,999' '8,399'
 '7,299' '61,999' '12,099' '18,999' '7,499' '5,549' '20,999' '9,499'
 '1,299' '5,999' '8,499' '7,749' '6,999' '6,499' '13,999' '16,999'
 '10,499' '19,999' '15,999' '8,219' '11,499' '8,349' '7,990' '8,599'
 '11,980' '12,499' '2,049' '2,099' '2,199' '2,149' '14,490' '8,598'
 '25,999' '27,999' '10,299' '13,499' '8,999' '2,133' '15,499' '999'
 '12,367' '18,890' '1,397' '1,659' '749' '1,475' '2,599' '16,227' '24,999'
 '14,999' '16,215' '15,990' '2,579' '32,999' '12,970' '6,345' '1,449'
 '21,699' '2,499' '18,564' '1,549' '71,999' '13,590' '13,788' '11,673'
 '938' '989'

In [24]:
bins = ['1000', '15000', '30000','50000']
labels = ['Buget phone','Smart Phone', 'Premium Phones']
df['Label_prices'] = pd.cut(df['Prices'], bins=bins, labels=labels)

Here we have created a bins for the price attribute.

**Bins are created as mentioned below:**

Price laying between **1000-15000** and their coressponding products under this are mentioned has **Budget phone**.

Price laying between **15000-30000** and their coressponding products under this are mentioned has **smart phone**.

Price laying between **30000-50000** and their coressponding products under this are mentioned has **Premium phone**.

And a seperate dummy column has been created to check whether the bins were properly implemented or not.

In [25]:
df.head(5)

Product Name  Prices  \
0     realme X50 Pro 5G (Rust Red, 256 GB)  47,999   
1  Xiaomi 12 Pro 5G (Couture Blue, 256 GB)  48,990   
2    Xiaomi 12 Pro 5G (Noir Black, 256 GB)  45,890   
3                            Good One 5026   1,099   
4  APPLE iPhone 8 Plus (Space Grey, 64 GB)  49,900   

                                         Description No. of Ratings  \
0  12 GB RAM | 256 GB ROM16.36 cm (6.44 inch) Ful...            778   
1  12 GB RAM | 256 GB ROM17.09 cm (6.73 inch) Ful...            177   
2  8 GB RAM | 256 GB ROM17.09 cm (6.73 inch) Full...            184   
3  32 MB RAM | 32 MB ROM | Expandable Upto 32 GB4...          1,127   
4  64 GB ROM13.97 cm (5.5 inch) Retina HD Display...          3,438   

  Label_Reviews    Label_prices  
0          High  Premium Phones  
1          High  Premium Phones  
2          High  Premium Phones  
3          High             NaN  
4          High  Premium Phones

In [26]:
print(df['Label_prices'].unique())

['Premium Phones', NaN, 'Smart Phone', 'Buget phone']
Categories (3, object): ['Buget phone' < 'Smart Phone' < 'Premium Phones']


In [27]:
df.isnull().sum()

Product Name        0
Prices              0
Description         0
No. of Ratings      0
Label_Reviews       0
Label_prices      487
dtype: int64

In [28]:
df=df.dropna()

In [29]:
df.count()

Product Name      444
Prices            444
Description       444
No. of Ratings    444
Label_Reviews     444
Label_prices      444
dtype: int64

In [30]:
df.isnull().sum()

Product Name      0
Prices            0
Description       0
No. of Ratings    0
Label_Reviews     0
Label_prices      0
dtype: int64

**Importing the preprocessed data into CSV file** 

In [31]:
df.to_csv("Flipcart_Preprocessed_Data_final.csv",index=False)

Here we have converted the processed the data into csv file in order to use this for our further process of data analysis in the name of Flipcart_Preprocessed_Data_final.csv

# Inserting csv file into mangodb server

In [32]:
!pip install "pymongo[srv]"

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 492.1/492.1 kB 16.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 283.7/283.7 kB 13.4 MB/s eta 0:00:00


Inserting preprocessed data into mongodb server
For this we are using **pymongo** python package

Here we are creating **client** instance and connected to server.

By using **client** instance we are sending ping command for successful connection and printing user defined message ,if connection is unsuccessful we are printing the exception message 

In [33]:
import pymongo
uri = "mongodb+srv://anup:dbs123@webscrapping.mg2ri2a.mongodb.net/?retryWrites=true&w=majority"
# Create a new client and connect to the server
client = pymongo.MongoClient(uri)
# Send a ping to confirm a successful connection
try:
    client.admin.command('ping')
    print("Pinged your deployment. You successfully connected to MongoDB!")
except Exception as e:
    print(e)

Pinged your deployment. You successfully connected to MongoDB!


Following code lines are used for creating **database** and **collection** name into the mongodb server using client instance

**Database**:Flipcart_Data

**Collection**:MobilePhones





In [34]:
db=client["Flipcart_Data"]
collection=db["MobilePhones"]

This line is used to print the list of available databases in the server

In [35]:
print(client.list_database_names())

['Flipcart_Data', 'sample_airbnb', 'sample_analytics', 'sample_geospatial', 'sample_guides', 'sample_mflix', 'sample_restaurants', 'sample_supplies', 'sample_training', 'sample_weatherdata', 'admin', 'local']


Below are the follwing steps and functions used to insert data into database

Import **csv** package to read csv file

There is no need to call file.close() when using **with** statement

**open()** function opens a file, and returns it as a file object.

Read the data from the csv file using **DictReader**

Create static **headers**

Drop the **collection** if already exists in server

Loop through data from **reader** and **headers** 

Store into the **JSON string row**

Print the **row**

Insert into the **collection** using **insert_one()**










In [36]:
import csv
with open('Flipcart_Preprocessed_Data_final.csv', 'r') as csvfile:
   reader = csv.DictReader( csvfile )
   headers= [ "Product Name", "Label_prices","Description","No. of Ratings","Label_Reviews","Prices"]
   collection.drop()
   for each in reader:
    row={}
    for field in headers:
        row[field]=each[field]
    print (row)
    collection.insert_one(row)
print('Data inserted successfully')

{'Product Name': 'realme X50 Pro 5G (Rust Red, 256 GB)', 'Label_prices': 'Premium Phones', 'Description': '12 GB RAM | 256 GB ROM16.36 cm (6.44 inch) Full HD+ Display64MP + 12MP + 8MP + 2MP | 32MP + 8MP Dual Front Camera4200 mAh Battery5G PhoneQualcomm Snapdragon 865 Processor65W Superdart ChargeBrand Warranty of 1 Year Available for Mobile and 6 Months for Accessories', 'No. of Ratings': '778', 'Label_Reviews': 'High', 'Prices': '47,999'}
{'Product Name': 'Xiaomi 12 Pro 5G (Couture Blue, 256 GB)', 'Label_prices': 'Premium Phones', 'Description': '12 GB RAM | 256 GB ROM17.09 cm (6.73 inch) Full HD+ AMOLED Display50MP Rear Camera4600 mAh BatterySnapdragon@ 8 Gen 1 Processor12 Months', 'No. of Ratings': '177', 'Label_Reviews': 'High', 'Prices': '48,990'}
{'Product Name': 'Xiaomi 12 Pro 5G (Noir Black, 256 GB)', 'Label_prices': 'Premium Phones', 'Description': '8 GB RAM | 256 GB ROM17.09 cm (6.73 inch) Full HD+ AMOLED Display50MP Rear Camera4600 mAh BatterySnapdragon@ 8 Gen 1 Processor12 

In [37]:
df = pd.read_csv("Flipcart_Preprocessed_Data_final.csv")

In [38]:
df.head()

Product Name  Prices  \
0     realme X50 Pro 5G (Rust Red, 256 GB)  47,999   
1  Xiaomi 12 Pro 5G (Couture Blue, 256 GB)  48,990   
2    Xiaomi 12 Pro 5G (Noir Black, 256 GB)  45,890   
3  APPLE iPhone 8 Plus (Space Grey, 64 GB)  49,900   
4              IQOO 9T 5G (LEGEND, 128 GB)  49,499   

                                         Description No. of Ratings  \
0  12 GB RAM | 256 GB ROM16.36 cm (6.44 inch) Ful...            778   
1  12 GB RAM | 256 GB ROM17.09 cm (6.73 inch) Ful...            177   
2  8 GB RAM | 256 GB ROM17.09 cm (6.73 inch) Full...            184   
3  64 GB ROM13.97 cm (5.5 inch) Retina HD Display...          3,438   
4  8 GB RAM | 128 GB ROM17.22 cm (6.78 inch) Disp...             54   

  Label_Reviews    Label_prices  
0          High  Premium Phones  
1          High  Premium Phones  
2          High  Premium Phones  
3          High  Premium Phones  
4          High  Premium Phones

**QUESTION1:** *What are the names of the products in the dataset that have received high or medium ratings?*

In [39]:
high_reviews_products = df[df["Label_Reviews"] == "High"]

# Display the product names along with their corresponding reviews
print(high_reviews_products[["Product Name", "Label_Reviews","Prices"]])

print('high_reviews_products')

                                     Product Name Label_Reviews  Prices
0            realme X50 Pro 5G (Rust Red, 256 GB)          High  47,999
1         Xiaomi 12 Pro 5G (Couture Blue, 256 GB)          High  48,990
2           Xiaomi 12 Pro 5G (Noir Black, 256 GB)          High  45,890
3         APPLE iPhone 8 Plus (Space Grey, 64 GB)          High  49,900
4                     IQOO 9T 5G (LEGEND, 128 GB)          High  49,499
..                                            ...           ...     ...
439      REDMI Note 12 5G (Mystique Blue, 128 GB)          High  17,990
440             realme GT 2 (Paper White, 128 GB)          High  27,999
441             realme GT 2 (Paper Green, 128 GB)          High  27,999
442           REDMI Note 9 (Shadow Black, 128 GB)          High  14,999
443  realme Narzo 50 Pro 5G (Hyper Black, 128 GB)          High  20,966

[436 rows x 3 columns]
high_reviews_products


In [40]:
medium_reviews_products = df[df["Label_Reviews"] == "Medium"]

# Display the product names along with their corresponding reviews
print(medium_reviews_products[["Product Name", "Label_Reviews","Prices"]])

print('medium_reviews_products')

                                      Product Name Label_Reviews  Prices
7                   IQOO 9 Pro 5G (Legend, 256 GB)        Medium  48,990
8              IQOO 9 Pro 5G (Dark Cruise, 256 GB)        Medium  48,990
233  SAMSUNG Galaxy M33 5G (Emerald Brown, 128 GB)        Medium  18,636
293               vivo T1 44W (Starry Sky, 128 GB)        Medium  17,999
308      REDMI Note 12 Pro 5G (Onyx Black, 128 GB)        Medium  24,999
310            OPPO A78 5G (Glowing Black, 128 GB)        Medium  18,510
325                                      LAVA Flip        Medium   2,199
424                 realme 9i (Prism Blue, 128 GB)        Medium  15,999
medium_reviews_products


We can see that **436 products have received high reviews**, while **8 products have received average ratings**. Nevertheless, we didn't attempt to retrieve the names of the items with low ratings in this case since, as we can see from the unique counts of the variables in the label reviews section, there are only two entries, either High or Medium, and this leads us to the conclusion that no product in the dataset has an underrated review, which is a low rating.


**QUESTION2:** *What are the different categories of phones present in the data and what specific models fall under each category?*

In [41]:
Budget_products = df[df["Label_prices"] == "Buget phone"]

# Display the product names along with their corresponding reviews
print(Budget_products[["Product Name", "Label_prices","Prices"]])

print('Budget_products')

                                    Product Name Label_prices  Prices
104  SAMSUNG Galaxy F13 (Waterfall Blue, 128 GB)  Buget phone  11,999
105                 POCO M4 Pro (Yellow, 128 GB)  Buget phone  12,999
106        MOTOROLA g52 (Metallic White, 128 GB)  Buget phone  12,999
107          MOTOROLA G32 (Mineral Gray, 128 GB)  Buget phone  12,099
115          MOTOROLA G32 (Satin Silver, 128 GB)  Buget phone  12,099
..                                           ...          ...     ...
427                realme 9i (Prism Blue, 64 GB)  Buget phone  13,499
428               realme 9i (Prism Black, 64 GB)  Buget phone  13,499
429   Infinix Hot 12 Pro (Electric Blue, 128 GB)  Buget phone  12,999
435            realme 9 5G (Meteor Black, 64 GB)  Buget phone  15,999
442          REDMI Note 9 (Shadow Black, 128 GB)  Buget phone  14,999

[141 rows x 3 columns]
Budget_products


We can see that there are **141 goods in the budget phone category**. **The price range for budget phones in this dataset is about between 1,000 and 15,999.**

In [42]:
smart_products = df[df["Label_prices"] == "Smart Phone"]

# Display the product names along with their corresponding reviews
print(smart_products[["Product Name", "Label_prices","Prices"]])

print('smart_products')

                                          Product Name Label_prices  Prices
21   Xiaomi 11T Pro 5G Hyperphone (Celestial Magic,...  Smart Phone  29,990
42      SAMSUNG Galaxy A34 5G (Awesome Silver, 128 GB)  Smart Phone  30,999
43        SAMSUNG Galaxy A34 5G (Awesome Lime, 128 GB)  Smart Phone  30,999
46                vivo V23 5G (Stardust Black, 256 GB)  Smart Phone  29,999
77                       Mi 10T (Cosmic Black, 128 GB)  Smart Phone  26,399
..                                                 ...          ...     ...
438                      SAMSUNG Guru Music 2 SM-B315E  Smart Phone   2,022
439           REDMI Note 12 5G (Mystique Blue, 128 GB)  Smart Phone  17,990
440                  realme GT 2 (Paper White, 128 GB)  Smart Phone  27,999
441                  realme GT 2 (Paper Green, 128 GB)  Smart Phone  27,999
443       realme Narzo 50 Pro 5G (Hyper Black, 128 GB)  Smart Phone  20,966

[176 rows x 3 columns]
smart_products


It is clear from the output above that there are **176 goods in the smart phone** **category**, with prices for these products ranging from **15,999 to 30,999.**

In [43]:
premium_products = df[df["Label_prices"] == "Premium Phones"]

# Display the product names along with their corresponding reviews
print(premium_products[["Product Name", "Label_prices","Prices"]])

print('premium_products')

                                     Product Name    Label_prices  Prices
0            realme X50 Pro 5G (Rust Red, 256 GB)  Premium Phones  47,999
1         Xiaomi 12 Pro 5G (Couture Blue, 256 GB)  Premium Phones  48,990
2           Xiaomi 12 Pro 5G (Noir Black, 256 GB)  Premium Phones  45,890
3         APPLE iPhone 8 Plus (Space Grey, 64 GB)  Premium Phones  49,900
4                     IQOO 9T 5G (LEGEND, 128 GB)  Premium Phones  49,499
..                                            ...             ...     ...
374  REDMI Note 12 Pro+ 5G (Arctic White, 256 GB)  Premium Phones  32,999
399                  Nokia C01 Plus (Blue, 16 GB)  Premium Phones   5,180
402               itel A27 (Silver Purple, 32 GB)  Premium Phones   5,098
413             realme GT 2 (Paper White, 256 GB)  Premium Phones  31,999
426                    itel A60 (DawnBlue, 32 GB)  Premium Phones   5,996

[127 rows x 3 columns]
premium_products


We can see from the result above that there are**127 goods in the premium phone category**, with premium phone prices in this dataset ranging from **31000 to 50000**.

**QUESTION3:** *What are the products that have a total number of rating greater than 500, and what are their respective reviews and product names?*

In [44]:
df = df.replace(',', '', regex=True)

In [45]:
df.head()

Product Name Prices  \
0     realme X50 Pro 5G (Rust Red 256 GB)  47999   
1  Xiaomi 12 Pro 5G (Couture Blue 256 GB)  48990   
2    Xiaomi 12 Pro 5G (Noir Black 256 GB)  45890   
3  APPLE iPhone 8 Plus (Space Grey 64 GB)  49900   
4              IQOO 9T 5G (LEGEND 128 GB)  49499   

                                         Description No. of Ratings  \
0  12 GB RAM | 256 GB ROM16.36 cm (6.44 inch) Ful...            778   
1  12 GB RAM | 256 GB ROM17.09 cm (6.73 inch) Ful...            177   
2  8 GB RAM | 256 GB ROM17.09 cm (6.73 inch) Full...            184   
3  64 GB ROM13.97 cm (5.5 inch) Retina HD Display...           3438   
4  8 GB RAM | 128 GB ROM17.22 cm (6.78 inch) Disp...             54   

  Label_Reviews    Label_prices  
0          High  Premium Phones  
1          High  Premium Phones  
2          High  Premium Phones  
3          High  Premium Phones  
4          High  Premium Phones

In [46]:
print(df['No. of Ratings'].unique())

['778' '177' '184' '3438' '54' '2380' '4' '206' '190930' '461' '42711'
 '1239' '99' '509' '544' '1519' '106' '21' '8696' '159' '266' '1085'
 '11062' '623' '3773' '1086' '2977' '249' '3050' '1127' '3815' '6542'
 '328' '4469' '496' '857' '3652' '193' '4704' '4022' '1895' '145542' '599'
 '218' '818' '244' '100737' '7961' '6150' '21939' '11370' '18184' '214122'
 '65594' '137729' '11572' '216530' '201465' '5039' '3609' '46161' '164621'
 '3255' '9619' '33264' '29003' '8985' '15416' '1107' '50' '179240' '18368'
 '28918' '35101' '1213' '2275' '25085' '16894' '1587' '40547' '5230'
 '36878' '8288' '141693' '3303' '1425' '2027' '3833' '959' '202228' '4861'
 '56990' '1574' '30450' '9043' '17680' '44890' '12809' '3147' '15902'
 '4151' '5830' '15343' '6070' '4361' '57032' '5501' '66408' '22176'
 '23321' '15580' '8326' '1755' '2791' '2393' '1817' '5244' '3195' '2032'
 '16999' '57364' '1231' '6189' '100541' '463' '7625' '29930' '3247' '1707'
 '13599' '7888' '235499' '208548' '2055' '120' '866' '13266'

In [47]:
# Convert the "total number of ratings" column to integer values
df["No. of Ratings"] = df["No. of Ratings"].astype(int)

# Find the products with higher total number of ratings
high_rated_products = df[df["No. of Ratings"] > 500]

# Display the output with the corresponding product name and ratings
for index, row in high_rated_products.iterrows():
    print(f"Product Name: {row['Product Name']}\nRating: {row['Label_Reviews']}\nNo. of Ratings: {row['No. of Ratings']}\nPrice:{row['Prices']}\n")


Product Name: realme X50 Pro 5G (Rust Red 256 GB)
Rating: High
No. of Ratings: 778
Price:47999

Product Name: APPLE iPhone 8 Plus (Space Grey 64 GB)
Rating: High
No. of Ratings: 3438
Price:49900

Product Name: OPPO Reno8 Pro 5G (Glazed Black 256 GB)
Rating: High
No. of Ratings: 2380
Price:45999

Product Name: APPLE iPhone 8 Plus (Gold 64 GB)
Rating: High
No. of Ratings: 3438
Price:49900

Product Name: APPLE iPhone 8 Plus (Silver 64 GB)
Rating: High
No. of Ratings: 3438
Price:49900

Product Name: APPLE iPhone 11 (White 128 GB)
Rating: High
No. of Ratings: 190930
Price:44999

Product Name: APPLE iPhone 11 (Black 128 GB)
Rating: High
No. of Ratings: 190930
Price:44999

Product Name: SAMSUNG Galaxy S21 FE 5G (Olive 128 GB)
Rating: High
No. of Ratings: 42711
Price:34999

Product Name: APPLE iPhone 11 (White 64 GB)
Rating: High
No. of Ratings: 190930
Price:40999

Product Name: APPLE iPhone 11 (Black 64 GB)
Rating: High
No. of Ratings: 190930
Price:40999

Product Name: OnePlus 10R 5G (Forest 

In [48]:
# Convert the "total number of ratings" column to integer values
df["No. of Ratings"] = df["No. of Ratings"].astype(int)

# Find the products with higher total number of ratings
high_rated_products = df[df["No. of Ratings"] > 500]

# Create a new DataFrame with the relevant columns
output_df = high_rated_products[["Product Name", "Label_Reviews", "No. of Ratings","Prices"]]

# Sort the output DataFrame in ascending order based on "No. of Ratings"
output_df = output_df.sort_values("No. of Ratings", ascending=False)

# Display the output in a formatted manner
print(output_df.to_string(index=False))

                                                         Product Name Label_Reviews  No. of Ratings Prices
                                 MOTOROLA g72 (Meteorite Grey 128 GB)          High          297962  15999
              IQOO Z6 Lite 5G (Without Charger) (Mystic Night 128 GB)          High          235499  16799
                              REDMI 11 Prime 5G (Meadow Green 128 GB)          High          235499  16670
                                 MOTOROLA g52 (Metallic White 128 GB)          High          216530  12999
                                      vivo T2 5G (Nitro Blaze 128 GB)          High          214122  18999
                               ASUS ZenFone 5Z (Midnight Blue 128 GB)          High          214122  36299
                                   MOTOROLA G32 (Mineral Gray 128 GB)          High          214122  12099
                               OPPO Reno3 Pro (Midnight Black 256 GB)          High          214122  35990
                                     

In this section, we can see all of the products with more than 500 ratings, as well as the product name, price, and ratings for that specific product.

**QUESTION4:** *What are the most commonly mentioned features in the descriptions of the phones?*

In [49]:
pip install wordcloud matplotlib


Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [50]:
import nltk
nltk.download('punkt')
from nltk.corpus import stopwords
nltk.download('stopwords')



[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

In [51]:
import nltk
from nltk.tokenize import word_tokenize
from nltk.probability import FreqDist

# Concatenate all the descriptions into a single string
text = ' '.join(df['Description'].values)

# Tokenize the text into individual words
tokens = word_tokenize(text)

# Remove stop words and punctuation
stopwords = nltk.corpus.stopwords.words('english')
filtered_tokens = [word.lower() for word in tokens if word.lower() not in stopwords and word.isalpha()]

# Compute the frequency distribution of the tokens
fdist = FreqDist(filtered_tokens)

# Print the 15 most common features
print(fdist.most_common(15))



[('gb', 824), ('warranty', 450), ('cm', 444), ('inch', 444), ('mah', 426), ('ram', 420), ('year', 386), ('months', 355), ('front', 335), ('accessories', 314), ('full', 269), ('manufacturer', 207), ('upto', 199), ('expandable', 196), ('rom', 195)]


We wanted to see what common key features were identified in the descriptions of each product in this analysis so that we might contemplate how buyers try to obtain information to make decisions.

We can observe that most items include the words like 'GB of the memory either it is **RAM or ROM, 'warranty for the product', 'display size', 'manufacturer',  'battery backup mAh', ' front camera and so on**, which helps buyers make informed decisions.


Because this is a Machine Learning topic and we are taking the same course, we thought it would be fun to run nltk (Natural Language Toolkit) on our dataset.

**QUESTION5:**  What is the distribution of ratings across the phones in the dataset and also show with any plots.

In [57]:
import plotly.express as px
# Create a histogram of the ratings, colored by product name
fig = px.histogram(df, x="No. of Ratings", nbins=10, color="Product Name", title="Distribution of Ratings by Product Name")

# Show the plot
fig.show()


After preprocessing the dataset, we got **444 products**. To understand the distribution of ratings for the products, we attempted to utilize histogram for visual depiction.

In Question 4, the most number of ratings for a product is 2,49,942, and the lowest is 502. Based on this data.


According to the plot, nearly 400 products have ratings ranging from 0 to 50,000, and a few devices have received more evaluations from customers and can be classified as popular phones.

**QUESTION6:** *What is the range of prices for each category of phones, and how do they compare with each other?*

In [58]:
# Create a box plot of prices for each category of phones
# Convert the 'Prices' column to a numerical data type
df['Prices'] = df['Prices'].astype(float)

# Create an interactive box plot of prices for each category of phones
fig = px.box(df, x='Label_prices', y='Prices', color='Label_prices')

# Set the axis labels and title
fig.update_xaxes(title_text='Price Category')
fig.update_yaxes(title_text='Price')
fig.update_layout(title='Distribution of Prices by Category')

# Show the plot
fig.show()



We utilized a box plot to determine the price distribution of the categories. 

Because the dataset we have is for less than 50,000 INR, the basic pricing of Premium phones ranges from 31,000 INR to 49,999 INR. 


Smart phone costs range from 16,000 INR to 30,999 INR, while Budget phone prices range from 11,249 INR to 15,999 INR.

We can see from the dataset that there are a few outliers in the 'Premium Phones' and 'Smart Phone' categories that are excluded from the fences because those devices do not fall within the price bracket of the category defined.

**QUESTION7:** What is the percentage breakdown of reviews across different categories (high, medium, low)?

In [59]:
# Count the number of reviews for each category
counts = df['Label_Reviews'].value_counts().reset_index()

# Create an interactive pie chart of the review distribution
fig = px.pie(counts, values='Label_Reviews', names='index', 
             title='Distribution of Reviews', 
             labels={'index': 'Review Category', 'Reviews': 'Count'})

# Show the plot
fig.show()


We considered utilizing an interactive pie chart to provide the answer. It is possible to see two slices, each of which represents a review category (High & Medium), and whose size reflects the proportion of High & Medium reviews. Consequently, making a pie chart gave users a quick and simple way to compare the proportion of ratings in each category.

**QUESTION8:** *What is the distribution of the total number of ratings for different phone categories (budget, smart, premium) based on their review ratings (high, medium, low)?*

In [60]:
# Create a pivot table to aggregate the number of ratings by category and review
ratings = pd.pivot_table(df, values='No. of Ratings', index='Label_prices', columns='Label_Reviews')

# Reset the index to create separate columns for each review category
ratings = ratings.reset_index()

# Melt the ratings DataFrame to create a long-form DataFrame for the stacked bar chart
ratings_melt = pd.melt(ratings, id_vars='Label_prices', var_name='Label_Reviews', value_name='No. of Ratings')

# Create a stacked bar chart
fig = px.bar(ratings_melt, x='Label_prices', y='No. of Ratings', color='Label_Reviews', barmode='stack')

# Set the title and axis labels
fig.update_layout(title='Number of Ratings by Category and Review', xaxis_title='Category',
                  yaxis_title='Total Number of Ratings')

# Show the plot
fig.show()


We intended to address this problem by producing a histogram; we see a peak in the frequency of ratings in the premium Phones category, indicating that there are more people who favor premium phones. The histogram also reveals that fewer users ranked the mobile phones as medium, implying that medium ratings are less common in the sample.

**QUESTION9:** How do the average prices of budget, smart, and premium phones vary based on their review ratings (high, medium, low)?

In [61]:
# Calculate the average prices by category and review
avg_prices = df.groupby(['Label_prices', 'Label_Reviews']).mean()[['Prices']]

# Reset the index to create separate columns for each review category
avg_prices = avg_prices.reset_index()

# Create a comparison bar chart
fig = px.bar(avg_prices, x='Label_prices', y='Prices', color='Label_Reviews', barmode='group')

# Set the title and axis labels
fig.update_layout(title='Average Prices by Category and Review', xaxis_title='Category',
                  yaxis_title='Average Price')

# Show the plot
fig.show()


<ipython-input-61-3e04e768774d>:2: FutureWarning:

The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.



Based on the ratings, we are attempting to determine the average price for each category. The average cost of a Premium phone with a high or medium rating is 33,387 INR and 20,062 INR, respectively. The average cost of a budget phone with high and medium ratings is 13,736 INR and 14,732 INR, respectively. The average price of a smart phone with a high or medium rating is 19,658 INR and 15,624 INR, respectively.

It can be seen that, although being pricey, Premium phones have received great evaluations due to the capabilities they provide to users. Budget phones, on the other hand, have received more middling ratings despite giving low prices, although users are not as satisfied as with Premium phones.